src/util/system.cpp

```diff
@@ -214,30 +210,43 @@ static util::SettingsValue InterpretOption(std::string& section, std::string& ke
     }
     if (key.substr(0, 2) == "no") {
         key.erase(0, 2);
-        // Double negatives like -nofoo=0 are supported (but discouraged)
-        if (!InterpretBool(value)) {
-            LogPrintf("Warning: parsed potentially confusing double-negative -%s=%s\n", key, value);
-            return true;
-        }
         return false;
     }
-    return value;
+    return true;
 }
 
 /**
- * Check settings value validity according to flags.
+ * Interpret settings value based on registered flags.
  *
- * TODO: Add more meaningful error checks here in the future
- * See "here's how the flags are meant to behave" in
- * https://github.com/bitcoin/bitcoin/pull/16097#issuecomment-514627823
+ * @param[in]   key      name of settings key, only used for error messages
+ * @param[in]   value    string value of setting to be parsed
+ * @param[in]   negated  whether to treat setting as negated
+ * @param[in]   flags    ArgsManager registered argument flags
+ * @param[out]  error    Error description if settings value is not valid
+ *
+ * @return parsed settings value if it is valid, otherwise nullopt accompanied
+ * by a descriptive error string
  */
-static bool CheckValid(const std::string& key, const util::SettingsValue& val, unsigned int flags, std::string& error)
-{
-    if (val.isBool() && !(flags & ArgsManager::ALLOW_BOOL)) {
-        error = strprintf("Negating of -%s is meaningless and therefore forbidden", key);
+static std::optional<util::SettingsValue> InterpretValue(const std::string& key,
+    const std::string& value,
+    bool negated,
+    unsigned int flags,
+    std::string& error)
```

Comment:
nit: to make `clang-format` happy:
```suggestion
static std::optional<util::SettingsValue> InterpretValue(
    const std::string& key,
    const std::string& value,
    bool negated,
    unsigned int flags,
    std::string& error)
```